pydantic/_internal/_fields.py

```diff
@@ -79,3 +90,162 @@ def _update_attrs(self, constraints: dict[str, Any], attrs: set[str] | None = No
             setattr(self, k, v)
+
+
+def collect_fields(  # noqa: C901
+    cls: type[Any],
+    bases: tuple[type[Any], ...],
+    types_namespace: dict[str, Any] | None,
+    *,
+    typevars_map: dict[str, Any] | None = None,
+    is_dataclass: bool = False,
+    dc_kw_only: bool | None = None,
+) -> dict[str, FieldInfo]:
+    """
+    Collect the fields of a pydantic model or dataclass, also returns the model/class ref to use in the
+    core-schema.
+
+    :param cls: BaseModel or dataclass
+    :param bases: parents of the class, generally `cls.__bases__`
+    :param types_namespace: optional extra namespace to look for types in
+    :param typevars_map: TODO ???
+    :param is_dataclass: whether the class is a dataclass, used to decide about kw_only setting
+    :param dc_kw_only: whether the whole dataclass is kw_only
+    """
+    from ..fields import FieldInfo
+
+    module_name = getattr(cls, '__module__', None)
+    global_ns: dict[str, Any] | None = None
+    if module_name:
+        try:
+            global_ns = sys.modules[module_name].__dict__
+        except KeyError:
+            # happens occasionally, see https://github.com/pydantic/pydantic/issues/2363
+            pass
+
+    # get type hints and raise a PydanticUndefinedAnnotation if any types are undefined
+    try:
+        type_hints = get_type_hints(cls, global_ns, types_namespace, include_extras=True)
+    except NameError as e:
+        try:
+            name = e.name
+        except AttributeError:
+            m = re.search(r".*'(.+?)'", str(e))
+            if m:
+                name = m.group(1)
+            else:
+                # should never happen
+                raise
+        raise PydanticUndefinedAnnotation(name) from e
+
+    # https://docs.python.org/3/howto/annotations.html#accessing-the-annotations-dict-of-an-object-in-python-3-9-and-older
+    # annotations is only used for finding fields in parent classes
+    annotations = cls.__dict__.get('__annotations__', {})
+    fields: dict[str, FieldInfo] = {}
+
+    # FIXME (dataclasses) what are we doing here?
+    # cls_fields: dict[str, FieldInfo] = getattr(cls, 'model_fields', None) or getattr(cls, '__pydantic_fields__', {})
+
+    # currently just used for `init=False` dataclass fields, but could be used more
+    omitted_fields: set[str] | None = getattr(cls, '__pydantic_omitted_fields__', None)
+
+    for ann_name, ann_type in type_hints.items():
+        if ann_name.startswith('_') or is_classvar(ann_type) or (omitted_fields and ann_name in omitted_fields):
+            continue
+
+        # raise a PydanticUndefinedAnnotation if type is undefined
+        if isinstance(ann_type, ForwardRef):
+            raise PydanticUndefinedAnnotation(ann_type.__forward_arg__)
+
+        # FIXME (dataclasses) what are we doing here?
+        pydantic_generic_typevars_map = getattr(cls, '__pydantic_generic_typevars_map__', None)
```

Comment:
It looks to me like this may be a bug (if it's still present on main), since the typevars map is applied to everything below.

Reply:
can I let you check that once this is merged?

Reply:
This was indeed a bug, fixed in #5209, with an appropriate test added.